File: backend/src/main/java/com/spareparts/inventory/entity/Product.java

package com.spareparts.inventory.entity;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Table(name = "products")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Product {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "wholesaler_id", nullable = false)
    private User wholesaler;

    @NotBlank
    @Size(max = 255)
    private String name;

    @NotBlank
    @Size(max = 100)
    @Column(unique = true)
    private String partNumber;

    @Size(max = 50)
    private String rackNumber;

    @NotNull
    @Column(columnDefinition = "numeric(38,2) default 0.0")
    private BigDecimal mrp = BigDecimal.ZERO;

    @NotNull
    @Column(columnDefinition = "numeric(38,2) default 0.0")
    private BigDecimal sellingPrice = BigDecimal.ZERO;

    @NotNull
    @Column(columnDefinition = "numeric(38,2) default 0.0")
    private BigDecimal wholesalerPrice = BigDecimal.ZERO;

    @NotNull
    @Column(columnDefinition = "numeric(38,2) default 0.0")
    private BigDecimal retailerPrice = BigDecimal.ZERO;

    @NotNull
    @Column(columnDefinition = "numeric(38,2) default 0.0")
    private BigDecimal mechanicPrice = BigDecimal.ZERO;

    @NotNull
    @Column(columnDefinition = "integer default 0")
    private Integer stock = 0;

    @NotNull
    @Column(columnDefinition = "boolean default true")
    private boolean enabled = true;

    @Column(columnDefinition = "integer default 1")
    private Integer minOrderQty = 1;

    @Column(name = "is_featured", columnDefinition = "boolean default false")
    private boolean isFeatured = false;

    @ElementCollection
    @CollectionTable(name = "product_images", joinColumns = @JoinColumn(name = "product_id"))
    @Column(name = "image_link")
    private java.util.List<String> imageLinks = new java.util.ArrayList<>();

    @Size(max = 500)
    private String imagePath;

    @Size(max = 500)
    private String imageLink;

    @Column(columnDefinition = "TEXT")
    private String description;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id")
    private Category category;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @Column(name = "deleted", nullable = false, columnDefinition = "boolean default false")
    private boolean deleted = false;

    @Enumerated(EnumType.STRING)
    @Column(name = "offer_type", length = 20)
    private OfferType offerType = OfferType.NONE;

    @Column(name = "offer_min_qty", columnDefinition = "integer")
    private Integer offerMinQty;

    public enum OfferType {
        NONE, DAILY, WEEKLY
    }
}

File: spare_parts_app_new/lib/models/product.dart
class Product {
  final int id;
  final String name;
  final String partNumber;
  final String? rackNumber;
  final double mrp;
  final double sellingPrice; // General/Default price
  final double wholesalerPrice;
  final double retailerPrice;
  final double mechanicPrice;
  final int stock;
  final int wholesalerId;
  final String? imagePath;
  final String? imageLink;
  final String? description;
  final bool enabled;
  final int? categoryId;
  final String? categoryName;
  final String? categoryImagePath;
  final String? categoryImageLink;

  Product({
    required this.id,
    required this.name,
    required this.partNumber,
    this.rackNumber,
    required this.mrp,
    required this.sellingPrice,
    required this.wholesalerPrice,
    required this.retailerPrice,
    required this.mechanicPrice,
    required this.stock,
    required this.wholesalerId,
    this.imagePath,
    this.imageLink,
    this.description,
    this.enabled = true,
    this.categoryId,
    this.categoryName,
    this.categoryImagePath,
    this.categoryImageLink,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id'] ?? 0,
      name: json['name'] ?? '',
      partNumber: json['partNumber'] ?? '',
      rackNumber: json['rackNumber'],
      mrp: (json['mrp'] as num? ?? 0).toDouble(),
      sellingPrice: (json['sellingPrice'] as num? ?? 0).toDouble(),
      wholesalerPrice:
          (json['wholesalerPrice'] as num? ?? json['sellingPrice'] as num? ?? 0)
              .toDouble(),
      retailerPrice:
          (json['retailerPrice'] as num? ?? json['sellingPrice'] as num? ?? 0)
              .toDouble(),
      mechanicPrice:
          (json['mechanicPrice'] as num? ?? json['sellingPrice'] as num? ?? 0)
              .toDouble(),
      stock: json['stock'] ?? 0,
      wholesalerId: json['wholesalerId'] ?? 0,
      imagePath: json['imagePath'],
      imageLink: json['imageLink'],
      description: json['description'],
      enabled: json['enabled'] is bool
          ? json['enabled']
          : (json['enabled'] ?? 1) == 1,
      categoryId: json['categoryId'] is int
          ? json['categoryId']
          : (json['categoryId'] as num?)?.toInt(),
      categoryName: json['categoryName'],
      categoryImagePath: json['categoryImagePath'],
      categoryImageLink: json['categoryImageLink'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'partNumber': partNumber,
      'rackNumber': rackNumber,
      'mrp': mrp,
      'sellingPrice': sellingPrice,
      'wholesalerPrice': wholesalerPrice,
      'retailerPrice': retailerPrice,
      'mechanicPrice': mechanicPrice,
      'stock': stock,
      'wholesalerId': wholesalerId,
      'imagePath': imagePath,
      'imageLink': imageLink,
      'description': description,
      'enabled': enabled ? 1 : 0,
      'categoryId': categoryId,
      'categoryName': categoryName,
      'categoryImagePath': categoryImagePath,
      'categoryImageLink': categoryImageLink,
    };
  }

  Product copyWith({
    int? id,
    String? name,
    String? partNumber,
    String? rackNumber,
    double? mrp,
    double? sellingPrice,
    double? wholesalerPrice,
    double? retailerPrice,
    double? mechanicPrice,
    int? stock,
    int? wholesalerId,
    String? imagePath,
    String? description,
    bool? enabled,
    int? categoryId,
    String? categoryName,
  }) {
    return Product(
      id: id ?? this.id,
      name: name ?? this.name,
      partNumber: partNumber ?? this.partNumber,
      rackNumber: rackNumber ?? this.rackNumber,
      mrp: mrp ?? this.mrp,
      sellingPrice: sellingPrice ?? this.sellingPrice,
      wholesalerPrice: wholesalerPrice ?? this.wholesalerPrice,
      retailerPrice: retailerPrice ?? this.retailerPrice,
      mechanicPrice: mechanicPrice ?? this.mechanicPrice,
      stock: stock ?? this.stock,
      wholesalerId: wholesalerId ?? this.wholesalerId,
      imagePath: imagePath ?? this.imagePath,
      description: description ?? this.description,
      enabled: enabled ?? this.enabled,
      categoryId: categoryId ?? this.categoryId,
      categoryName: categoryName ?? this.categoryName,
    );
  }
}
